materialize-starburst/starburst.go

```diff
@@ -0,0 +1,430 @@
+type config struct {
+	Host               string `json:"host" jsonschema:"title=Host and optional port" jsonschema_extras:"order=0"`
+	Catalog            string `json:"catalog" jsonschema:"title=Catalog" jsonschema_extras:"order=1"`
+	Schema             string `json:"schema" jsonschema:"title=Schema" jsonschema_extras:"order=2"`
+	Account            string `json:"account" jsonschema:"title=Account" jsonschema_extras:"order=3"`
+	Password           string `json:"password" jsonschema:"title=Password" jsonschema_extras:"secret=true,order=4"`
+	AWSAccessKeyID     string `json:"awsAccessKeyId" jsonschema:"title=Access Key ID" jsonschema_extras:"order=5"`
+	AWSSecretAccessKey string `json:"awsSecretAccessKey" jsonschema:"title=Secret Access Key" jsonschema_extras:"secret=true,order=6"`
+	Region             string `json:"region" jsonschema:"title=Region" jsonschema_extras:"order=7"`
+	Bucket             string `json:"bucket" jsonschema:"title=Bucket" jsonschema_extras:"order=8"`
+	DataStoreDir       string `json:"dataStoreDir" jsonschema:"title=Directory for storing temporary files" jsonschema_extras:"order=8"`
+
+	Advanced advancedConfig `json:"advanced,omitempty" jsonschema:"title=Advanced Options,description=Options for advanced users. You should not typically need to modify these." jsonschema_extras:"advanced=true"`
+}
+
+type advancedConfig struct {
+	UpdateDelay string `json:"updateDelay,omitempty" jsonschema:"title=Update Delay,description=Potentially reduce active warehouse time by increasing the delay between updates. Defaults to 30 minutes if unset.,enum=0s,enum=15m,enum=30m,enum=1h,enum=2h,enum=4h"`
+}
+
+// ToURI converts the Config to a DSN string.
+func (c *config) ToURI() string {
+
+	var params = make(url.Values)
+	params.Add("catalog", c.Catalog)
+	params.Add("schema", c.Schema)
+
+	var uri = url.URL{
+		Scheme:   "https",
+		Host:     c.Host,
+		User:     url.UserPassword(c.Account, c.Password),
+		RawQuery: params.Encode(),
+	}
+
+	return uri.String()
+}
+
+func (c *config) Validate() error {
+	var requiredProperties = [][]string{
+		{"account", c.Account},
+		{"host", c.Host},
+		{"catalog", c.Catalog},
+		{"schema", c.Schema},
+		{"password", c.Password},
+		{"awsAccessKeyId", c.AWSAccessKeyID},
+		{"awsSecretAccessKey", c.AWSSecretAccessKey},
+		{"region", c.Region},
+		{"bucket", c.Bucket},
+		{"dataStoreDir", c.DataStoreDir},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+
+	return nil
+}
+
+type tableConfig struct {
+	Table  string `json:"table" jsonschema:"title=Table,description=Name of the table" jsonschema_extras:"x-collection-name=true"`
+	Schema string `json:"schema,omitempty" jsonschema:"title=Schema,description=Schema where the table resides"`
+	Delta  bool   `json:"delta_updates,omitempty" jsonschema:"default=false,title=Delta Update,description=Should updates to this table be done via delta updates. Default is false."`
```

Comment:
> If there's not a way to do an idempotent `COPY INTO` or something like that

@williamhbaker I'd be interested to know about why we need `COPY INTO`.

The current logic copies the files to a "directory" path on the object storage (same as `materialize-databricks) and queries the _Hive_ external table having the location of the object storage.
I don't see why we need `COPY INTO`.

Please detail a bit why/ whether is idempotency needed in the context of this connector.

Reply:
Typically delta updates will append updates to a table exclusively with "insert" type of actions (often something like a `COPY INTO`, depending on the system), and never update anything in place. A primary key constraint is not enforced, since it is expected that there be multiple rows for a given key.

For example, if we had these documents in a collection with the key `key`, and each document was materialized in a separate transaction:
```json
{"key":"first","value":1}
{"key":"first","value":2}
{"key":"first","value":3}
```

...a delta updates materialization might produce a table like this:
```
+-------+-------+
|  key  | value |
+-------+-------+
| first |     1 |
| first |     2 |
| first |     3 |
+-------+-------+
```

But, if we are running only merge queries which match on the value of the collection key, we'd get a table like this:
```
+-------+-------+
|  key  | value |
+-------+-------+
| first |     3 |
+-------+-------+
```

There's a bit more about delta updates in the [Estuary docs](https://docs.estuary.dev/concepts/materialization/#delta-updates)

That all said, there's no reason the connector absolutely _must_ support delta updates, so it would be fine to not have that as an option and only do standard updates.

Reply:
Thank you @williamhbaker for the explanation.

I went through the Esturay docs and understand what _Delta Updates_ is.
What I'm missing from the docs is when would a materialization be inclined to have _Delta Updates_ instead of letting this job for the _Flow_ ?


I've went through Databricks' [documentation](https://docs.databricks.com/en/sql/language-manual/delta-copy-into.html) for `COPY INTO` and read the "idempotent" bumper sticker in the description of the command.
I'm guessing though that this is linked to the fact that there is a kind of tracking to which files have been copied into the destination table and which haven't.
I'm not sure however how does `COPY INTO` help to enable _Delta Updates_ for the Databricks materialization.
https://github.com/estuary/connectors/blob/main/materialize-databricks/driver.go#L37

I'd appreciate having this insight.
Thank you very much for taking the time to explain the concept to me. I'm still interested to dig deeper :)

Reply:
When using delta updates, we use `COPY INTO` to directly load the rows into the target table, and don't use a temporary table. So, `COPY INTO` being idempotent is necessary, because the connector may re-run the `COPY INTO` query in a recovery situation upon restarting, and if it weren't idempotent it would result in duplicated rows.

For a contrasting example, if we used a temporary table for delta updates, we would need to not run a merge query, but rather something like `INSERT INTO <target_table> SELECT * FROM <temp_table>`, and that would not be idempotent, since it would result in duplicated rows if the connector had to re-run the query.

> What I'm missing from the docs is when would a materialization be inclined to have Delta Updates instead of letting this job for the Flow ?

The delta updates mode of operation for a materialization is getting at how documents are reduced into the target table. The Flow runtime will actually always fully reduce documents within each transaction, so there is never more than one instance of each key in a transaction. With delta updates though, documents are _not_ reduced in the destination system...existing rows are never updated. As a result, no keys will be loaded from the target table when running in delta updates - there's no reason to reduce with existing rows. And, all of the documents to store will be considered "new": see an example [here](https://github.com/estuary/connectors/blob/68cf3d58db0095ee1eff12fc78294ad59fd51e7a/materialize-snowflake/snowflake.go#L469-L471) for how this is used to determine if a merge query vs. `COPY INTO` is used for `materialize-snowflake`.

Users may want to enable delta updates for a materialization if they know that their source data never has any updates, which can make the materialization work a little faster since it won't have to load anything or ever do any updates in the destination. It can also enables an approximation of a "history mode", which shows all changes to a key. But again, supporting delta updates isn't strictly necessary for a connector, and I don't think it would be a problem to just not have that as an option for now if it can't be supported.

Reply:
I removed option for delta updates.